### ISSUE TITLE
Retrieving the response body as a List of POJOs fails with `RestClient` but passes with `WebTestClient`

### ISSUE THREAD

**Issue by donalmurtagh** [spring-projects/spring-framework#31574]
**Affects:** spring 6.1.0-RC2 with spring boot 3.2.0-RC2 and JDK21

**Description**
I've attached a sample Spring Boot application with a single endpoint that returns the response as a wrapped `List<Label>`. When I retrieve the response with `WebTestClient` the list (correctly) contains `Label` instances, but when I retrieve the response with `RestClient` the list contains `Map` instances.

In both cases, I've used the same `ParameterizedTypeReference` when retrieving the response

**Steps to Reproduce**
[demo.zip](https://github.com/spring-projects/spring-framework/files/13300191/demo.zip)

The application contains 2 integration tests, one of which invokes the endpoint with a `RestClient` and the other uses `WebTestClient`. The latter test passes, but the former fails.

**Comment by sdeleuze** [comment]
Likely caused by the fact `contextClass` should be `null` for client side use case (it is currently set to `bodyClass`).

**Comment by donalmurtagh** [comment]
> Likely caused by the fact `contextClass` should be `null` for client side use case (it is currently set to `bodyClass`).

@sdeleuze thanks for the reply, but I'm not sure what exactly you think is the problem is. Could you provide a bit more detail? I guess the problem is either in the way I'm constructing the `RestClient`

```java
@PostConstruct
private void buildRestClient() {
    restClient = restClientBuilder.baseUrl("http://localhost:" + port)
        .defaultHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE)
        .build();
}
```

or maybe there's something wrong with the way I'm using it to invoke the endpoint?

```java
restClient.get()
        .uri("/api/labels")
        .retrieve()
        .toEntity(new ParameterizedTypeReference<ResponseWrapper<List<Label>>>() { })
        .getBody()
        .data();
```

**Comment by donalmurtagh** [comment]
@sdeleuze thanks for the fix. I've confirmed that this is resolved in the latest 6.1.0-SNAPSHOT
